### DISCUSSION TITLE
Fault simulation

### DISCUSSION THREAD

**Discussion by Blebowski** [nickg/nvc#895] [Ideas]
Hi,
just capturing an idea. @nickg as you are now working towards
Verilog implementation, it is likely that NVC will be able to parse a verilog
netlist soon (netlists usually contain only structural Verilog, so very small sub-set of the language).
It would be nice to try to simulate an input test pattern (e.g. in VCD or STIL),
and  measure fault coverage obtained by such pattern.

**Comment by Blebowski** [comment]
Hi @nickg,

I was wondering if this is something you would be open to.

The goals of fault simulation typically are:
- Measure fault coverage of DUT / chip / device.
- Examine behavior of a device in environment where SEU (Single event upstets) may occur.

Having at least the first one would be a major thing in an open source tool.

* Measure DUTs fault coverage *

When an ASIC is being manufactured, having high fault coverage is important to get small amount of defective parts in the field. The fault coverage can be obtained from ATPG (for automated test patterns), or by fault-simulation.

During fault simulation, a tool simulates DUT two times:
- Gold simulation
- Faulty simulation

In the gold simulation, the simulator applies the stimuly (either test-bench around a DUT, or VCD input),
and:
- Records the DUTs outputs (in case of TB used as stimuli)
- Compares with VCDs output (in case of VCD based stimuli) -> Must match for VCD stimuli to be consistent.

Then in a faulty simulation, the simulator applies the same stimuli to DUTs input / inout pins, and compares
the output / inout pins with those of "gold simulation". If there is a mismatch somewhere, it means that the
fault was detected.

Then the simulator loops through all faults in the device, and fault-simulates each one of them. The
fault coverage is basically the amount of detected vs. undetected faults.

There are different fault models, the basic "stuck-at" fault model would be a good place to start.
Basically, the simulator assumes that the DUT is a gate-level netlist, and each input or output of
a PDK cell can contain stuck-at-0 and stuck-at-1 fault. The cell is typically recognized by being a module
defined under `celldefine` directive.

Obviously there are more elaborate options possible, but this simple approach would be a good place
to start.

**Reply by nickg** [reply]
I think this kind of tool would be best implemented on top of VHPI/VPI rather than coupling it to the main simulator. It could be bundled in the `contrib/` directory if it's useful to a sufficient number of users.

**Reply by kev-cam** [reply]
@Blebowski 

I'm going to be doing it because I have this patent that needs a simulator -

https://patents.google.com/patent/US20230334213A1/en?inventor=D+Kevin+Cameron

- it's a hierarchical approach that came out of working on IEEE P2427 (and watching it deliver nothing useful).

**Comment by Blebowski** [comment]
Yeah, that is a good idea. I think the commercial tools use similar approach, e.g. waveform viewer communicates with simulator via VPI / VHPI to query the design hierarchy.

Looking at the VPI definition in System Verilog LRM, there is `vpiCellInstance` to figure out module is a cell.

**Comment by Blebowski** [comment]
A related discussion I was starting in verilator:
https://github.com/verilator/verilator/issues/7098

**Comment by byzantic** [comment]
I don't really think there is a good case for trying to adapt nvc for fault simulation. Trying to implement it through VHPI or VPI would only be suitable for injecting a limited number of faults; it might be useful for teaching purposes, but not for practical designs. 

For an engineering flow, what you actually need is ATPG. Modern designs are too large to test using purely functional tests, which is where fault simulators were originally used. Instea, scan-testing is used. Since the entire design then becomes a flat combinatorial circuit (with input and output flip-flops), it is simpler to deal with, and there are some open-source tools that are available for ATPG. OpenRoad has some support.

**Reply by Blebowski** [reply]
Hi @byzantic ,

sure, but for smaller designs (e.g. less than 10 milion gates), the fault simulation is still often used. Also, sometimes part of the design are not scanned, so you need to test them by functional tests. In such case, you need to combine the faults generated by ATPG with the ones tested by Fault simulation.

The typical industry fault simulation tools employ "concurrent fault simulation" that speeds the fault simulation up. My approach with NVC would be having sequential simulation, but simulate multiple simulations in parallel. So, trade compute power for expensive SW license.

There are additional use-cases for fault simulation such as functional safety (SEU resilience) or Security (laser attack resistance) where there are VERY expensive tools that do the fault simulation.

My idea would be to have "wrapper" tool for NVC (E.g. sth like `nvc_fault`) that would use VHPI / VPI to simulate individual faults, and would provide ways how to e.g. custom clasify a fault to be "detected" (e.g. if output of some alarm fires) - rather than just comparing DUT outputs with gold simulation.

**Reply by byzantic** [reply]
OK, good luck! I think injecting faults one by one is going to be very slow, though.

**Reply by Blebowski** [reply]
Thanks, I agree that the scaling simply by compute power has limits. At some point, there would be need for better algorithmic solution.

**Reply by kev-cam** [reply]
I did some work with atalanta for ATPG to create block-level models for working and defective behavior. The methodology is hierarchical, so you don't end up trying to do stuck-at faults on wires in billion transistor designs.
